README.md
# rpi-surveillance
Make a surveillance camera from your Raspberry Pi 4!

The surveillance is built as following: the camera records 10 seconds video 
and if a motion was detected - sends the video to telegram channel.

The timestamp is printed on videos, so it is better to set the correct time on 
your Raspberry Pi.

The motion detection works in the following way: the camera’s H.264 encoder 
calculates motion vector estimates while generating compressed video. Using 
these vectors we threshold them by `--magnitude-th` argument. If more than 
`--vectors-quorum` vectors thresholded - mark current frame as containing 
motion. If there are more than `--detection-frames` consecutive frames with 
motion - motion detected.

Tested on Raspberry Pi 4 (4 RAM) + NoIR Camera V2.

## Installation

### Install package

Install Python 3 requirements:
```shell
pip3 install --user -r requirements.txt
```

Install provided `.deb` package:
```shell
sudo dpkg -i <path/to/downloaded/rpi-surveillance.deb>
sudo apt install -f
```

Note: the installation supposes that you already enabled camera module on your Raspberry Pi.

### Create telegram bot and chat

1. Write to `@BotFather` in telegram and create a bot:
```
/start
/newbot
<name of your bot>
<username of your bot>_bot
```
You will get the TOKEN. Save it for future use.

2. Create a private channel where you will receive video sequences with motion.
3. Add created bot to the channel (rerquires only "post messages" permission).
4. Send message `test` to the channel.
5. Run `/usr/lib/rpi-surveillance/get_channel_id` to get the CHANNEL_ID. 
   Save it for future use.


## Usage

To launch surveillance just run `rpi-surveillance` with your TOKEN and 
CHANNEL_ID, for example:
```shell
rpi-surveillance --token 1259140266:WAaqkMycra87ECzRZwa6Z_8T9KB4N-8OPI --channel-id -1003209177928
```

You can set various parameters of the surveillance:
```
usage: rpi-surveillance [-h] [--config CONFIG] --token TOKEN --channel-id
                        CHANNEL_ID [--temp-dir TEMP_DIR] [--log-file LOG_FILE]
                        [--resolution {640x480,1280x720,1920x1080}]
                        [--fps {25,30,60}] [--rotation {0,90,180,270}]
                        [--duration DURATION] [--magnitude-th MAGNITUDE_TH]
                        [--vectors-quorum VECTORS_QUORUM]
                        [--detection-frames DETECTION_FRAMES]

optional arguments:
  -h, --help            show this help message and exit
  --config CONFIG       Path to config file.
  --token TOKEN         Token for your telegram bot.
  --channel-id CHANNEL_ID
                        Telegram channel ID. If you don't have it please, send
                        a message to your channel and run /usr/lib/rpi-
                        surveillance/get_channel_id with your token.
  --temp-dir TEMP_DIR   Path to temporary directory for video saving before
                        sending to channel. Don't change it if you don't know
                        what you're doing.
  --log-file LOG_FILE   Path to log file for logging.
  --resolution {640x480,1280x720,1920x1080}
                        Camera resolution. Default - 640x480.
  --fps {25,30,60}      Frames per second. Default - 25.
  --rotation {0,90,180,270}
                        Frame rotation. Default - 0.
  --duration DURATION   Duration of videos in seconds. Default - 10.
  --magnitude-th MAGNITUDE_TH
                        Magnitude threshold for motion detection (lower - more
                        sensitive). Defaults: for 640x480 - 15, for 1280x720 -
                        40, for 1920x1080 - 65.
  --vectors-quorum VECTORS_QUORUM
                        Vectors quorum for motion detection (lower - more
                        sensitive). Defaults: for 640x480 - 10, for 1280x720 -
                        20, for 1920x1080 - 40.
  --detection-frames DETECTION_FRAMES
                        The number of consecutive frames with detected motion
                        to send an alert.
```

## Build
Build was done using `dpkg-buildpackage`.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Unreleased
Unreleased changes will be added to this section first.


## [2.2.2] - 2021-12-15
### Added
- Added a telegram notification when the app has been launched.

[2.2.2]: https://github.com/resolator/rpi-surveillance/compare/v2.2.1...v2.2.2


## [2.2.1] - 2021-12-07
### Added
- Added camera release when closing the application.

[2.2.1]: https://github.com/resolator/rpi-surveillance/compare/v2.2.0...v2.2.1


## [2.2.0] - 2021-07-07
### Added
- Additional argument `--detection-frames` for motion detection.

### Changed
- Motion detection improved for cases with camera noise.

[2.2.0]: https://github.com/resolator/rpi-surveillance/compare/v2.1.0...v2.2.0


## [2.1.0] - 2021-07-07
### Added
- Motion detection argumets.

### Removed
- Unused telegram pooling.

[2.1.0]: https://github.com/resolator/rpi-surveillance/compare/v2.0.1...v2.1.0


## [2.0.1] - 2021-06-29
### Changed
- Temporary directory mount rights.

[2.0.1]: https://github.com/resolator/rpi-surveillance/compare/v2.0.0...v2.0.1


## [2.0.0] - 2021-06-29
### Added
- Temporary directory cleanup when starting the work.

### Changed
- Distribution way (developed debian package).
- Moved tmpfs creation to the installation stage.

[2.0.0]: https://github.com/resolator/rpi-surveillance/compare/v1.0.0...v2.0.0


## [1.0.0] - 2021-06-21
Initial release.

[1.0.0]: https://github.com/resolator/rpi-surveillance/releases/tag/v1.0.0


install.sh
#!/usr/bin/env bash

HELP_MESSAGE=$(cat <<-END
usage: install.sh [--help] TOKEN CHANNEL_ID

positional arguments:
  TOKEN    Telegram token of your bot.
  CHANNEL_ID  Channel ID where your bot is added and can send messages.

optional arguments:
  --help            Show this help message and exit.
END
)

# check arguments
TOKEN=$1
if [ "${TOKEN}" == "" ]; then
    echo "[ERROR] Please enter the Telegram token for your bot."
    exit 1
elif [ "${TOKEN}" == "--help" ]; then
    echo "${HELP_MESSAGE}"
    exit 0
fi

CHANNEL_ID=$2
if [ "${CHANNEL_ID}" == "" ]; then
    echo "[ERROR] Please enter the Telegram channel where your bot is."
    exit 2
fi

# create default config file
REPO_DIR=$(dirname "$0")
CONFIG_FILE="${REPO_DIR}/default.cfg"

if [ -f "${CONFIG_FILE}" ]; then
    echo "[ERROR] ${CONFIG_FILE} exists. Please, remove it before continuing."
    exit 3
fi

# install requirements
sudo apt update
sudo apt install python3-pip python3-venv libatlas-base-dev gpac

# create and setup venv
VENV_DIR="${REPO_DIR}/.venv"
python3 -m venv "${VENV_DIR}"
source "${VENV_DIR}/bin/activate"
pip3 install --upgrade pip
pip3 install -r "${REPO_DIR}/requirements.txt"
deactivate

# create default config file
touch "${CONFIG_FILE}"
echo "token=${TOKEN}" >> "${CONFIG_FILE}"
echo "channel-id=${CHANNEL_ID}" >> "${CONFIG_FILE}"
echo "temp-dir=/tmp/rpi-surveillance" >> "${CONFIG_FILE}"
echo "resolution=640x480" >> "${CONFIG_FILE}"
echo "fps=25" >> "${CONFIG_FILE}"
echo "rotation=0" >> "${CONFIG_FILE}"
echo "duration=10" >> "${CONFIG_FILE}"

echo -e "\nDone! We created ${CONFIG_FILE} with default configuration that you can change."
echo "Now you can run it using start.sh"


get_chat_id.sh
#!/usr/bin/env bash

HELP_MESSAGE=$(cat <<-END
usage: install.sh [--help] TOKEN

positional arguments:
  TOKEN    Telegram token of your bot.

optional arguments:
  --help            Show this help message and exit.
END
)

# check argument
TOKEN=$1
if [ "${TOKEN}" == "" ]; then
    echo "[ERROR] Please enter the Telegram token for your bot."
    exit 1
elif [ "${TOKEN}" == "--help" ]; then
    echo "${HELP_MESSAGE}"
    exit 0
fi

# extract chat/channel ID
CHAT_ID=$(curl -s "https://api.telegram.org/bot${TOKEN}/getUpdates" | grep -oPm 1 "\"sender_chat\":{\"id\":\K[^<]+?(?=,)")
if [ "${CHAT_ID}" == "" ]; then
    echo -e "\n[WARNING] Could not parse chat ID. Please, check that:"
    echo "- you send a message \"test\" to the chat with bot"
    echo "- bot has access to messages in this chat"
    echo "- bot TOKEN is correct"
    exit 2
else
    echo -e "\nChat/channel ID: ${CHAT_ID}"
fi


surveillance.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*
import os
import logging
import threading
import configargparse

import numpy as np

from pathlib import Path
from telegram.ext import Updater
from datetime import datetime as dtm
from picamera.array import PiMotionAnalysis
from picamera import PiCamera, Color as PiColor
from logging.handlers import RotatingFileHandler


def get_args():
    """Arguments parser and checker."""
    parser = configargparse.ArgumentParser(description=__doc__)
    parser.add_argument('--config', is_config_file=True,
                        help='Path to config file.')
    parser.add_argument('--token', required=True,
                        help='Token for your telegram bot.')
    parser.add_argument('--channel-id', required=True,
                        help='Telegram channel ID.')
    parser.add_argument('--temp-dir', type=Path,
                        default=Path('/tmp/rpi-surveillance'),
                        help='Path to temporary directory for video saving '
                             'before sending to channel.')
    parser.add_argument('--resolution', default='640x480',
                        choices=['640x480', '1280x720', '1920x1080'],
                        help='Camera resolution.')
    parser.add_argument('--fps', type=int, default=30, choices=[25, 30, 60],
                        help='Frames per second.')
    parser.add_argument('--rotation', type=int, default=0,
                        choices=[0, 90, 180, 270],
                        help='Frame rotation.')
    parser.add_argument('--duration', type=int, default=5,
                        help='Duration of videos in seconds.')
    parser.add_argument('--log-file',
                        help='Path to log file for logging.')

    # check args
    args = parser.parse_args()
    assert args.duration > 2, 'duration is too low.'

    return args


class DetectMotion(PiMotionAnalysis):
    def __init__(self, camera, vectors_quorum=10, magnitude_th=60):
        super().__init__(camera)
        self.magnitude_th = magnitude_th
        self.vectors_quorum = vectors_quorum
        self.motion = False

    def analyze(self, motions):
        motions = np.sqrt(
            np.square(motions['x'].astype(np.float)) +
            np.square(motions['y'].astype(np.float))
        ).clip(0, 255).astype(np.uint8)

        if (motions > self.magnitude_th).sum() > self.vectors_quorum:
            self.motion = True


def send_record(bot, channel_id, h264_path, fps):
    """Convert a record to from h264 to mp4 and send it to a telegram channel.

    Parameters
    ----------
    bot : telegram.ext.Updater.bot
        Telegram bot instance.
    channel_id : str
        Telegram channel_id of your channel.
    h264_path : pathlib.Path
        Path to h264 record.
    fps : int
        Frames per seconds for h264 record.

    """
    # convert h264 to mp4
    mp4_path = h264_path.parent.joinpath(h264_path.stem + '.mp4')
    cmd = f'MP4Box -add {h264_path} -fps {fps} {mp4_path} >> /dev/null 2>&1'
    os.system(cmd)

    # send to telegram channel
    bot.send_video(chat_id=channel_id,
                   video=open(mp4_path, 'rb'),
                   supports_streaming=True)

    # clean up
    os.remove(mp4_path)
    os.remove(h264_path)


def main():
    """Application entry point."""
    args = get_args()

    # setup logging
    date_format = '%Y-%m-%d_%H-%M-%S'
    logging.basicConfig(level=logging.INFO,
                        format='[%(levelname)s] %(asctime)s - %(message)s',
                        datefmt=date_format)

    logger = logging.getLogger('rpi_logger')
    logger.info('Initializing')

    if args.log_file is not None:
        handler = RotatingFileHandler(args.log_file, 'w', 1024 * 1024 * 50, 5)
        logger.addHandler(handler)

    # connect to telegram bot
    updater = Updater(token=args.token)
    updater.start_polling()

    # setup camera
    camera = PiCamera(resolution=args.resolution, framerate=args.fps)
    camera.rotation = args.rotation
    camera.annotate_text_size = 14
    camera.annotate_background = PiColor('black')

    # setup move detection
    detection_data = {'640x480': (10, 20),
                      '1280x720': (20, 50),
                      '1920x1080': (40, 80)}
    vectors_quorum, magnitude_th = detection_data[args.resolution]
    output = DetectMotion(camera, vectors_quorum, magnitude_th)

    logger.info('Initialization completed')
    logger.info('Start recording')
    while True:
        start = dtm.now()
        h264_path = args.temp_dir.joinpath(
            start.strftime(date_format) + '.h264')

        output.motion = False
        camera.annotate_text = dtm.now().strftime(date_format)
        camera.start_recording(str(h264_path),
                               format='h264',
                               motion_output=output)

        # make a record
        while (dtm.now() - start).seconds < args.duration:
            camera.annotate_text = dtm.now().strftime(date_format)
            camera.wait_recording(0.2)

        # save file
        camera.stop_recording()

        # if there was a motion - convert and send video asynchronously
        if output.motion:
            logger.warning('Motion detected, sending a record')
            threading.Thread(
                target=send_record,
                args=[updater.bot, args.channel_id, h264_path, args.fps]
            ).start()
        else:
            os.remove(h264_path)


if __name__ == '__main__':
    main()


get_channel_id.sh
#!/usr/bin/env bash

HELP_MESSAGE=$(cat <<-END
usage: install.sh [--help] TOKEN

positional arguments:
  TOKEN    Telegram token of your bot.

optional arguments:
  --help            Show this help message and exit.
END
)

# check argument
TOKEN=$1
if [ "${TOKEN}" == "" ]; then
    echo "[ERROR] Please enter the Telegram token for your bot."
    exit 1
elif [ "${TOKEN}" == "--help" ]; then
    echo "${HELP_MESSAGE}"
    exit 0
fi

# extract channel ID
CHANNEL_ID=$(curl -s "https://api.telegram.org/bot${TOKEN}/getUpdates" | grep -oPm 1 "\"sender_chat\":{\"id\":\K[^<]+?(?=,)")
if [ "${CHANNEL_ID}" == "" ]; then
    echo -e "\n[WARNING] Could not parse channel ID. Please, check that:"
    echo "- you send a message \"test\" to the channel with bot"
    echo "- bot has access to messages in this channel"
    echo "- bot TOKEN is correct"
    exit 2
else
    echo -e "\nCHANNEL_ID: ${CHANNEL_ID}"
fi


video_capturer.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*
import os
import logging
import argparse
import threading

import numpy as np

from pathlib import Path
from telegram.ext import Updater
from datetime import datetime as dtm
from picamera.array import PiMotionAnalysis
from picamera import PiCamera, Color as PiColor
from logging.handlers import RotatingFileHandler


def get_args():
    """Arguments parser and checker."""
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--token', required=True,
                        help='Token for your telegram bot.')
    parser.add_argument('--chat-id', required=True,
                        help='Telegram chat ID.')
    parser.add_argument('--temp-dir', type=Path,
                        default=Path('/tmp/rpi-surveillance'),
                        help='Path to temporary directory for video saving '
                             'before sending to chat.')
    parser.add_argument('--resolution', default='640x480',
                        choices=['640x480', '1280x720', '1920x1080'],
                        help='Camera resolution.')
    parser.add_argument('--fps', type=int, default=30, choices=[30, 60],
                        help='Frames per second.')
    parser.add_argument('--rotation', type=int, default=0,
                        choices=[0, 90, 180, 270],
                        help='Frame rotation.')
    parser.add_argument('--duration', type=int, default=5,
                        help='Duration of videos in seconds.')
    parser.add_argument('--log-file',
                        help='Path to log file for logging.')
    parser.add_argument('--save-to', type=Path,
                        help='Path to save dir.')

    # check args
    args = parser.parse_args()
    assert args.duration > 2, 'duration is too low.'

    return args


class DetectMotion(PiMotionAnalysis):
    def __init__(self, camera, vectors_quorum=10, magnitude_th=60):
        super().__init__(camera)
        self.magnitude_th = magnitude_th
        self.vectors_quorum = vectors_quorum
        self.motion = False

    def analyze(self, motions):
        motions = np.sqrt(
            np.square(motions['x'].astype(np.float)) +
            np.square(motions['y'].astype(np.float))
        ).clip(0, 255).astype(np.uint8)

        if (motions > self.magnitude_th).sum() > self.vectors_quorum:
            self.motion = True


def send_record(bot, chat_id, h264_path, fps):
    """Convert a record to from h264 to mp4 and send it to a telegram channel.

    Parameters
    ----------
    bot : telegram.ext.Updater.bot
        Telegram bot instance.
    chat_id : str
        Telegram chat id of your channel.
    h264_path : pathlib.Path
        Path to h264 record.
    fps : int
        Frames per seconds for h264 record.

    """
    # convert h264 to mp4
    mp4_path = h264_path.parent.joinpath(h264_path.stem + '.mp4')
    cmd = f'MP4Box -add {h264_path} -fps {fps} {mp4_path} >> /dev/null 2>&1'
    os.system(cmd)

    # send to telegram channel
    bot.send_video(chat_id=chat_id,
                   video=open(mp4_path, 'rb'),
                   supports_streaming=True)

    # clean up
    os.remove(mp4_path)
    os.remove(h264_path)


def main():
    """Application entry point."""
    args = get_args()

    # setup logging
    date_format = '%Y-%m-%d_%H-%M-%S'
    logging.basicConfig(level=logging.INFO,
                        format='[%(levelname)s] %(asctime)s - %(message)s',
                        datefmt=date_format)

    logger = logging.getLogger('rpi_logger')
    logger.info('Initializing')

    if args.log_file is not None:
        handler = RotatingFileHandler(args.log_file, 'w', 1024 * 1024 * 50, 5)
        logger.addHandler(handler)

    # connect to telegram bot
    updater = Updater(token=args.token)
    updater.start_polling()

    # setup camera
    camera = PiCamera(resolution=args.resolution, framerate=args.fps)
    camera.rotation = args.rotation
    camera.annotate_text_size = 14
    camera.annotate_background = PiColor('black')

    # setup move detection
    detection_data = {'640x480': (10, 20),
                      '1280x720': (15, 40),
                      '1920x1080': (20, 60)}
    vectors_quorum, magnitude_th = detection_data[args.resolution]
    output = DetectMotion(camera, vectors_quorum, magnitude_th)

    logger.info('Initialization completed')
    logger.info('Start recording')
    while True:
        start = dtm.now()
        h264_path = args.temp_dir.joinpath(
            start.strftime(date_format) + '.h264')

        output.motion = False
        camera.annotate_text = dtm.now().strftime(date_format)
        camera.start_recording(str(h264_path),
                               format='h264',
                               motion_output=output)

        # make a record
        while (dtm.now() - start).seconds < args.duration:
            camera.annotate_text = dtm.now().strftime(date_format)
            camera.wait_recording(0.2)

        # save file
        camera.stop_recording()

        # if there was a motion - convert and send video asynchronously
        if output.motion:
            logger.warning('Motion detected, sending a record')
            threading.Thread(
                target=send_record,
                args=[updater.bot, args.chat_id, h264_path, args.fps]
            ).start()
        else:
            os.remove(h264_path)


if __name__ == '__main__':
    main()


rpi_surveillance/app.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*
import os
import logging
import threading
import configargparse

import numpy as np

from pathlib import Path
from telegram.ext import Updater
from datetime import datetime as dtm
from picamera.array import PiMotionAnalysis
from picamera import PiCamera, Color as PiColor
from logging.handlers import RotatingFileHandler


def get_args():
    """Arguments parser and checker."""
    parser = configargparse.ArgumentParser(description=__doc__)
    parser.add_argument('--config', is_config_file=True,
                        help='Path to config file.')

    # telergam arguments
    parser.add_argument('--token', required=True,
                        help='Token for your telegram bot.')
    parser.add_argument('--channel-id', required=True,
                        help='Telegram channel ID. If you don\'t have it '
                             'please, send a message to your channel and '
                             'run /usr/lib/rpi-surveillance/get_channel_id '
                             'with your token.')

    # other arguments
    parser.add_argument('--temp-dir', type=Path,
                        default=Path('/usr/lib/rpi-surveillance/ram-dir'),
                        help='Path to temporary directory for video saving '
                             'before sending to channel. Don\'t change it if '
                             'you don\'t know what you\'re doing.')
    parser.add_argument('--log-file',
                        help='Path to log file for logging.')

    # camera arguments
    parser.add_argument('--resolution', default='640x480',
                        choices=['640x480', '1280x720', '1920x1080'],
                        help='Camera resolution. Default - 640x480.')
    parser.add_argument('--fps', type=int, default=25, choices=[25, 30, 60],
                        help='Frames per second. Default - 25.')
    parser.add_argument('--rotation', type=int, default=0,
                        choices=[0, 90, 180, 270],
                        help='Frame rotation. Default - 0.')
    parser.add_argument('--duration', type=int, default=10,
                        help='Duration of videos in seconds. Default - 10.')

    # detection arguments
    parser.add_argument('--magnitude-th', type=int,
                        help='Magnitude threshold for motion detection '
                             '(lower - more sensitive). '
                             'Defaults: for 640x480 - 15,'
                             '          for 1280x720 - 40,'
                             '          for 1920x1080 - 65.')
    parser.add_argument('--vectors-quorum', type=int,
                        help='Vectors quorum for motion detection '
                             '(lower - more sensitive). '
                             'Defaults: for 640x480 - 10,'
                             '          for 1280x720 - 20,'
                             '          for 1920x1080 - 40.')
    parser.add_argument('--detection-frames', type=int, default=4,
                        help='The number of consecutive frames with detected '
                             'motion to send an alert.')

    # check args
    args = parser.parse_args()
    assert args.duration > 2, 'duration is too low.'

    if args.magnitude_th is None:
        default_magnitudes = {'640x480': 15, '1280x720': 40, '1920x1080': 65}
        args.magnitude_th = default_magnitudes[args.resolution]

    if args.vectors_quorum is None:
        default_vectors = {'640x480': 10, '1280x720': 20, '1920x1080': 40}
        args.vectors_quorum = default_vectors[args.resolution]

    return args


class DetectMotion(PiMotionAnalysis):
    def __init__(self,
                 camera,
                 vectors_quorum=10,
                 magnitude_th=60,
                 motion_frames=4):
        super().__init__(camera)
        self._magnitude_th = magnitude_th
        self._vectors_quorum = vectors_quorum
        self._motion_frames = motion_frames
        self._motions_history = 0
        self._motion = False

    def analyze(self, motions):
        motions = np.sqrt(
            np.square(motions['x'].astype(np.float)) +
            np.square(motions['y'].astype(np.float))
        ).clip(0, 255).astype(np.uint8)

        # check motion on the current frame
        if (motions > self._magnitude_th).sum() > self._vectors_quorum:
            self._motions_history += 1
        else:
            self._motions_history = 0

        # check last N frames motion trigger
        if self._motions_history >= self._motion_frames:
            self._motion = True

    def reset_detection(self):
        """Clean detected motion."""
        self._motion = False

    def detect_motion(self):
        """Return True if motion was presented."""
        return self._motion


def send_record(bot, channel_id, h264_path, fps):
    """Convert a record to from h264 to mp4 and send it to a telegram channel.

    Parameters
    ----------
    bot : telegram.ext.Updater.bot
        Telegram bot instance.
    channel_id : str
        Telegram channel_id of your channel.
    h264_path : pathlib.Path
        Path to h264 record.
    fps : int
        Frames per seconds for h264 record.

    """
    # convert h264 to mp4
    mp4_path = h264_path.parent.joinpath(h264_path.stem + '.mp4')
    cmd = f'MP4Box -add {h264_path} -fps {fps} {mp4_path} >> /dev/null 2>&1'
    os.system(cmd)

    # send to telegram channel
    bot.send_video(chat_id=channel_id,
                   video=open(mp4_path, 'rb'),
                   supports_streaming=True)

    # clean up
    os.remove(mp4_path)
    os.remove(h264_path)


def main():
    """Application entry point."""
    args = get_args()

    # clean temp dir
    [os.remove(x) for x in args.temp_dir.glob('*.*')]

    # setup logging
    date_format = '%Y-%m-%d_%H-%M-%S'
    logging.basicConfig(level=logging.INFO,
                        format='[%(levelname)s] %(asctime)s - %(message)s',
                        datefmt=date_format)

    logger = logging.getLogger('rpi_logger')
    logger.info('Initializing')

    if args.log_file is not None:
        handler = RotatingFileHandler(args.log_file, 'w', 1024 * 1024 * 50, 5)
        logger.addHandler(handler)

    # connect to telegram bot
    updater = Updater(token=args.token)

    # setup camera
    camera = PiCamera(resolution=args.resolution, framerate=args.fps)
    camera.rotation = args.rotation
    camera.annotate_text_size = 14
    camera.annotate_background = PiColor('black')

    # setup move detection
    output = DetectMotion(camera,
                          args.vectors_quorum,
                          args.magnitude_th,
                          args.detection_frames)
    
    # send notification
    updater.bot.send_message(chat_id=args.channel_id,
                             text='Surveillance has been started.')
    
    logger.info('Initialization completed')
    logger.info('Start recording')
    try:
        while True:
            start = dtm.now()
            h264_path = args.temp_dir.joinpath(
                start.strftime(date_format) + '.h264')

            output.reset_detection()
            camera.annotate_text = dtm.now().strftime(date_format)
            camera.start_recording(str(h264_path),
                                format='h264',
                                motion_output=output)

            # make a record
            while (dtm.now() - start).seconds < args.duration:
                camera.annotate_text = dtm.now().strftime(date_format)
                camera.wait_recording(0.2)

            # save file
            camera.stop_recording()

            # if there was a motion - convert and send video asynchronously
            if output.detect_motion():
                logger.warning('Motion detected, sending a record')
                threading.Thread(
                    daemon=True,
                    target=send_record,
                    args=[updater.bot, args.channel_id, h264_path, args.fps]
                ).start()
            else:
                os.remove(h264_path)
    finally:
        camera.stop_recording()


if __name__ == '__main__':
    main()
